vantage/virtual_tag_config_resource_test.go

```diff
@@ -0,0 +1,174 @@
+			// Update: not specifying values
+			{
+				Config: testAccVantageVirtualTagConfigTf_basic(
+					"test",
+					keyV1,
+					!overridable,
+					backfillUntil,
+					"",
+				),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "key", keyV1),
+					resource.TestCheckResourceAttr(resourceName, "overridable", "false"),
+					resource.TestCheckResourceAttr(resourceName, "values.#", "0"),
+				),
+			},
+			// Update: set multiple values with filters
+			{
+				Config: testAccVantageVirtualTagConfigTf_basic(
+					"test",
+					keyV1,
+					!overridable,
+					backfillUntil,
+					tfValues(
+						[]map[string]string{
+							{
+								"name":   "value-0",
+								"filter": "(costs.provider = 'aws' AND costs.service = 'AmazonEC2') OR (costs.provider = 'gcp' AND costs.service = 'ComputeEngine')",
+							},
+							{
+								"name":   "value-1",
+								"filter": "(costs.provider = 'gcp' AND costs.service != 'ComputeEngine')",
+							},
+						},
+					),
+				),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "key", keyV1),
+					resource.TestCheckResourceAttr(resourceName, "overridable", "false"),
+					resource.TestCheckResourceAttr(resourceName, "values.#", "2"),
+					resource.TestCheckResourceAttr(resourceName, "values.0.name", "value-0"),
+					resource.TestCheckResourceAttr(resourceName, "values.1.name", "value-1"),
+				),
+			},
+			// Update: preserving existing values
+			{
+				Config: testAccVantageVirtualTagConfigTf_basic(
+					"test",
+					keyV1,
+					!overridable,
+					backfillUntil,
+					fmt.Sprintf(`values = %[1]s`, fromState(keyV1, "values")),
+				),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "key", keyV1),
+					resource.TestCheckResourceAttr(resourceName, "overridable", "false"),
+					resource.TestCheckResourceAttr(resourceName, "values.#", "2"),
+					resource.TestCheckResourceAttr(resourceName, "values.0.name", "value-0"),
+					resource.TestCheckResourceAttr(resourceName, "values.1.name", "value-1"),
+				),
+			},
+			// // Update: adding to existing values
+			// {
+			// 	Config: testAccVantageVirtualTagConfigTf_basic(
+			// 		"test",
+			// 		keyV1,
+			// 		!overridable,
+			// 		backfillUntil,
+			// 		fmt.Sprintf(`values = concat(%[1]s, [{ name = "value-2" }])`, fromState(keyV1, "values")),
+			// 	),
+			// 	Check: resource.ComposeAggregateTestCheckFunc(
+			// 		resource.TestCheckResourceAttr(resourceName, "key", keyV1),
+			// 		resource.TestCheckResourceAttr(resourceName, "overridable", "false"),
+			// 		resource.TestCheckResourceAttr(resourceName, "values.#", "3"),
+			// 		resource.TestCheckResourceAttr(resourceName, "values.0.name", "value-0"),
+			// 		resource.TestCheckResourceAttr(resourceName, "values.1.name", "value-1"),
+			// 		resource.TestCheckResourceAttr(resourceName, "values.2.name", "value-2"),
+			// 	),
+			// },
```

Comment:
@whereandy @macb (and anyone else with some terraform experience), what am I missing in this commented out test case?

When I attempt to use the state to specify an additional value, I get this from terraform calling out the new value when I apply:

```
virtual_tag_config_resource_test.go:44: Step 6/7 error: After applying this test step, the plan was not empty.    
    stdout:    
        
        
    Terraform used the selected providers to generate the following execution    
    plan. Resource actions are indicated with the following symbols:    
      ~ update in-place    
        
    Terraform will perform the following actions:    
        
      # vantage_virtual_tag_config.test will be updated in-place    
      ~ resource "vantage_virtual_tag_config" "test" {    
          ~ created_by_token = "usr_fc451e546aeaf056" -> (known after apply)    
          ~ values           = [    
              + {    
                  + filter = (known after apply)    
                  + name   = "value-2"    
                },    
                # (3 unchanged elements hidden)    
            ]    
            # (4 unchanged attributes hidden)    
        }    
        
    Plan: 0 to add, 1 to change, 0 to destroy.
```

Reply:
I have not used the terraform testing pieces at all. Does the resource work as expected against an API?

Reply:
It does, `values` is sent as [...previous, new value]

Reply:
hmm some stuff to noodle on brought up here: https://discuss.hashicorp.com/t/framework-migration-test-produces-non-empty-plan/54523/6

- custom plan checks ([link](https://developer.hashicorp.com/terraform/plugin/testing/acceptance-tests/plan-checks)) for debugging
- `PlanModifiers` ([link](https://developer.hashicorp.com/terraform/plugin/framework/resources/plan-modification))
- and implementing a custom type if it is (and it undoubtably is) a state comparison issue.
  I sidestepped using a custom type by just deserializing into an annotated struct for the values, but if I had the excuse of it helping here, it'd also clean up deserialization some.

Reply:
> It does, `values` is sent as [...previous, new value]

Ya, more meant after the apply, does the plan also reflect the expected values? ie. `terraform apply` followed by `terraform plan` show the expected no changes required.

Reply:
Oh hah. 😊
very reasonable question, but I've been keeping to the test suite so no idea, yet.

Reply:
next stop, that.

Reply:
I would start there tbh. My assumption is it will not return an empty plan because the state is not being updated as expected on this update case.

Reply:
The plan applies - I think it's an experience gap with me and the test suite. I'm going to leave things as-is and call it 'good enough' for now